### PR TITLE
Mandate X-Pack REST handler installed

### PR DESCRIPTION
**Pull request by BigPandaToo**
In order to provide a stronger guarantee to our solutions, that if a
cluster is running the default distribution and has security
(authentication) enabled,
then it will be provided by Elastic's security features, and users can
rely on it behaving in the ways they expect, this change mandate
that security in default distribution is provided by X-Pack by always
installing the Security Rest Filter.

Related: #70523

### CODE REVIEW

**Comment by BigPandaToo** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (tvernum) [inline server/src/main/java/org/elasticsearch/action/ActionModule.java:451]: Having a debug here seems weird to me. I feel like we should remove the check entirely, or make in a warning/error.
Maintaining this code, just to print a debug message that no one will see doesn't seem valuable.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1112]: I think we need this:
```suggestion
        final boolean ssl = enabled && HTTP_SSL_ENABLED.get(settings);
```
If security is disabled, then HTTP SSL will not be enabled either (even if the setting is true).

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:98]: Do we want this? I would have thought we'd keep the warning behaviour from #70114

**Review by tvernum — COMMENTED** [review]
We also need to fill in the PR description

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:98]: Right. I need to revisit this PR after I merge #701114

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by tvernum** [comment]
@BigPandaToo Can you please fill in the PR description.

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (tvernum) [inline server/src/main/java/org/elasticsearch/action/ActionModule.java:450]:
```suggestion
                    logger.warn("The " + plugin.getClass().getName() + " plugin installs a custom REST wrapper. This functionality " +
```

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (tvernum) [inline server/src/main/java/org/elasticsearch/action/ActionModule.java:451]: I don't think this wording is right - what does _should be disabled_ mean to the reader?

Take a step back - Who is going to read this message, and what action do you want them to take when they do?

The warning message should convey that information to them, as succinctly as it can.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1112]: This is the main part of the PR - it's the thing that actually achieves the result we're after, but I don't see anywhere where we test that this PR delivers the behaviour that we've decided on.

That is, we need a `testSecurityHandlerIsAlwaysInstalled`

**Review by tvernum — CHANGES_REQUESTED** [review]

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/action/ActionModule.java:449]: This check needs to be updated following the changes in #73245.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1124]: I think it reads better if we inline `httpSSLConfig`. Though it works in this case, assigning a variable to `null` and then deciding whether to use it based on another condition bother me.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:65]: Nit
```suggestion
        if (request.method() == Method.OPTIONS) {
```

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/action/ActionModule.java:451]: The code here only logs a warning when a custom plugin tries to install its own rest warpper. It doesn't really try to do anything to stop it though the message says "the plugin has been disabled"? From what I can tell, it would still go ahead and install the plugin and its rest wrapper?

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:548]: This does not really test that the rest wrapper is installed and is from the Security plugin. It only tests that a rest handler for path `/` is installed, which happens regardless with or without current changes.

In my opinion, a simple unit test for `Security#getRestHandlerWrapper` is sufficient, i.e. simply test the handler wrapper is returned regardless of the value of `xpack.security.enabled`.

If you prefer the current approach, it might be possible to register a new handler and try to invoke it. Along the way we can try inspect the call stack to ensure the security rest wrapper is in action. But in general, I feel testing at this level is not much, if at all, better than the unit test. If we do want to test the real thing, the same approach of inspecting call stack is probably better suited for an external cluster test.

**Review by ywangd — CHANGES_REQUESTED** [review]
Maybe I am missing something, but I don't think the code does what the warning message claims.

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (BigPandaToo) [reply server/src/main/java/org/elasticsearch/action/ActionModule.java]: It won't. Because of the change in `Security::getRestHandlerWrapper` ^^, xpack security rest handler will always be installed even when security is disabled. But you have a point here, this message will never be reached because we throw an exception two lines above if a second plugin is trying to install a rest handler. There are several options on how we can handle it:
1) Log the warning and throw (current implementation)
2) Log the warning, but only throw if it was an `es` plugin. Proceed ignoring 2nd rest handler if it was a 3rd party plugin
3) Log the warning and proceed ignoring 2nd rest handler regardless of the source of the plugin

I don't have strong preferences choosing between the options.  @bytebilly, what do you think?

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:547]: This test tests that despite `xpack.security.enabled` being set to `false`, the Rest Handler has been installed. But I, actually, added one more test which ensures the second plugin trying to install a rest handler will fail.

Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:547]: Aren't we trying to test whether the rest **wrapper** is installed instead of the rest "handler"?

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/ActionModule.java]: There is no guarantee that the Security plugin is the first ActionPlugin that installs a RestWrapper. We don't have any logic to ensure that. It is entirely possible that a custom plugin comes here first and installs the rest wrapper and in turn blocks the actual Security plugin.

When the above happens, the message will be logged. But it is misleading because nothing is disabled. The overall behaviour is more or less correct because the node will fail to launch (due to Security plugin trying to install the rest wrapper again). But I think the logic is confusing. I'd suggest: 
1. Throw an exception here and possibly remove the above exception throwing. 
2. Reword the message to make it clear a custom rest wrapper triggers an exception

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:547]: I think one implies another. There is no easy way to access the Wrapper through the rest controller, but we can tell the rest wrapper was installed by initializing rest handlers and trying to register another one.

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (bytebilly) [reply server/src/main/java/org/elasticsearch/action/ActionModule.java]: Could we check if the REST wrapper is being installed by the x-pack plugin, and throw a warning+exception otherwise? This should happen for each REST wrapper, including the first one.

If I understand them correctly, other proposals seem not hitting the target:

> Log the warning and throw (current implementation)
This doesn't ensure that the REST wrapper is from the x-pack plugin

> Log the warning, but only throw if it was an es plugin. Proceed ignoring 2nd rest handler if it was a 3rd party plugin
Not sure to get it, but we should avoid any kind of third-party plugin at all

> Log the warning and proceed ignoring 2nd rest handler regardless of the source of the plugin
This doesn't ensure that the REST wrapper is from the x-pack plugin

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:547]: > I think one implies another.

I disagree. This test passes even when no rest Wrapper is installed. That is, if you replace `Arrays.asList(xpackPlugin, security)` with `List.of()` in the constructor call of `ActionModule`, the test passes by throwing the same error.

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (BigPandaToo) [reply server/src/main/java/org/elasticsearch/action/ActionModule.java]: I actually though we do have the logic to ensure `es` plugind installed firs through plugin dependancy sorting... But now I see you are right.
Changing the logic to throw on any non `es` security plugin trying to install the handler.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java]: You are right! Changed the test to look at the debug message instead

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: Sorry but what I meant is to replace
```java
final SSLConfiguration httpSSLConfig = ssl ? getSslService().getHttpTransportSSLConfiguration() : null;        
boolean extractClientCertificate = ssl && getSslService().isSSLClientAuthEnabled(httpSSLConfig);
```

with 

```java
boolean extractClientCertificate = 
    ssl && getSslService().isSSLClientAuthEnabled(getSslService().getHttpTransportSSLConfiguration());
```

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:559]: This call is no longer needed since all necessary logic is within ActionModule's constructor.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:553]: Nit: This logging line is the first thing the code does when it sees a rest wrapper. So strictly speaking, at this point, it is still possible that the rest wrapper will be rejected later. It is not really an issue right now, because the code throws if the rest wrapper is not installed and the test would fail. But it does not reads too well for me.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:532]: I am not a fan of this test. As an unit test, I think it mixes two things:
* `Security` - Ensure a non-null restWrapper is always return by `getRestHandlerWrapper`.
* `ActionModule` - Ensure only restWrapper from `Security` can be installed

Each of the above can be tested independantly in a more unit-test-like fashion. And if we want to test the real thing, we can go for internal or external cluster. 

But I am happy to let it pass if Tim is OK with it.

**Review by ywangd — CHANGES_REQUESTED** [review]
I left a few more comments. Thanks!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:96]: I have discussed it with Fabio and he agreed that with Security On by Default we won't need it in 8.0. The change which is making the default behaviour to true is on it's way.

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:96]: As discussed, I am happy to let it pass if Tim is OK with the approach. Thanks!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1124]:
```suggestion
        boolean extractClientCertificate = false;
        if (enabled && HTTP_SSL_ENABLED.get(settings)) {
            final SSLConfiguration httpSSLConfig = getSslService().getHttpTransportSSLConfiguration();
            extractClientCertificate = getSslService().isSSLClientAuthEnabled(httpSSLConfig);
        }
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (tvernum) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:96]: I would prefer that we kept the warning. Removing features based on the premise that another change will come along and make them redundant is a risky business. If we change our plans for security on by default, then this warning might still be required, but it will have been lost.  
The likelihood is relatively low, but the safer option is to remove code when it's no longer needed rather than removing it when we decide it won't be needed in the future.

However, in this case, there's no simple way to keep the code around.
We need to know whether security is _implictly disabled_, and there's no method on `XPackLicenseState` for that, and I don't want to add one now.  
We can looking at `XPackSettings.SECURITY_ENABLED`, but then we need a copy of the node's `Settings`, which we don't have.

Because it's a warning, and not critical functionality, I think it's easier to just remove it, but that's not my preferred approach in general.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (tvernum) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:532]: Testing the ActionModule is tricky unless we change how it works. It has a hardcoded dependency on a particular classname, so it's impossible to mock it out (see https://github.com/elastic/elasticsearch/pull/73329).

We could soften that check to only check for the package name (`org.elasticsearch.xpack.security`), or to pass the name of the allowed plugin into the constructor. Either of those would then allow it to be tested in a more unit-test-like fashion. Otherwise the strict class-name dependency makes it difficult

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:600]: We shouldn't use an anonymous class for this. It means that the message is dependent on the synthentic name generated by the compiler, and that's too fragile to use.

Comment (BigPandaToo) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:600]: For the warning headers, we can as well wait for #71061 to be merged.

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/action/ActionModule.java:445]: Nit: I probably missed the discussion about why we prefer `getCanonicalName` over just `getName` here. But it seems to me that `getName` is more consistent since that is what we report in the error message (and also the debug log above)?

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (ywangd) [inline server/src/main/java/org/elasticsearch/action/ActionModule.java:447]: Nit: Since an exception is thrown immediately after this and will be logged. It seems this logging is no longer necessary.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:1126]: Nit: we can save the extra `finalExtractClientCertificate` variable.

```suggestion
        final boolean extractClientCertificate;
        if (enabled && HTTP_SSL_ENABLED.get(settings)) {
            final SSLConfiguration httpSSLConfig = getSslService().getHttpTransportSSLConfiguration();
            extractClientCertificate = getSslService().isSSLClientAuthEnabled(httpSSLConfig);
        } else {
            extractClientCertificate = false;
        }
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:65]: Nit: I'd prefer we keep the single outter if/else block. It reads better to me.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (ywangd) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:532]: The check is now relaxed to be package name. Do you think it is worth to have more unit test like tests? 
Technically, even with the previous check for exact class name, it is possible to have an unit test for `ActionModule`: Since `ActionModule` is in a different module and does not depend on xpack security, it is possible to create a stub `org.elasticsearch.xpack.security.Security` class in its test package and use that for the unit test.

**Review by ywangd — APPROVED** [review]
LGTM

A few nits and also I don't particular like the "unit test". But they don't need to block this PR. Thanks!

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:553]: I have added some comments to the test

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (tvernum) [inline server/src/main/java/org/elasticsearch/action/ActionModule.java:450]: Per https://github.com/elastic/elasticsearch/pull/73329, there should be tests for this functionality in ActionModuleTests.
Core shouldn't rely on x-pack to do its testing for it.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:571]: This belongs in `ActionModuleTests`, it's a testing the behaviour of `ActionModule`, so it shouldn't be here.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (tvernum) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:532]: I think it's testing 2 things, and I think it would be better if they were split into
1. `testSecurityPluginInstallsRestHandlerWrapperEvenIfSecurityIsDisabled` which can simply be a test that `Security.getRestHandlerWrapper()` returns non-null even when disabled.
2. `testSecurityRestHandlerWrapperCanBeInstalled()` which tests that ActionModule doesn't reject the wrapper installed by security (essentially this test).

File: server/src/main/java/org/elasticsearch/action/ActionModule.java
Comment (BigPandaToo) [reply server/src/main/java/org/elasticsearch/action/ActionModule.java:450]: Moved it to ActionModuleTests

File: server/src/test/java/org/elasticsearch/action/ActionModuleTests.java
Comment (tvernum) [inline server/src/test/java/org/elasticsearch/action/ActionModuleTests.java:241]: This is fine - there's no need to change it, but it would have been fine to just make this a no-op operator.
```suggestion
            return UnaryOperator.identity();
```
We really only care that the plugin returns _something_. We don't care what it does.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java
Comment (tvernum) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/SecurityTests.java:542]: Is this needed?
I think you can skip all the `ActionModule` part since this is a simple unit test on `Security`

**Review by tvernum — APPROVED** [review]
LGTM, but I have a couple of suggestions.

**Pull request merged**
